### DISCUSSION TITLE
Object not registered even though it exists

### DISCUSSION THREAD

**Discussion by AmbroiseJuston** [idaholab/moose#28004] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

I have a kernel called DiffusionCoupled. It is a custom kernel that I added to my application. I have both the .h and the .C files in their appropriate folder, include/kernels and src/kernels within the application (called difluoride). I have used this kernel in the past for previous applications, and everything seems to work fine when I use it in that application (called xenon). I then ported an exact copy of the kernel to the new app difluoride. I remember to change the name of the application within the .C kernel to re-register it to the new application. 

All this being done, I tried to run the application and I got the following error:

hwloc/linux: Ignoring PCI device with non-16bit domain.
Pass --enable-32bits-pci-domain to configure to support such devices
(warning: it would break the library ABI, don't enable unless really needed).

*** ERROR ***
A 'DiffusionCoupled' is not a registered object.

We loaded objects from the following libraries and still couldn't find your object:
        libphase_field-opt.la

If you are trying to find this object in a dynamically loaded library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.

Stack frames: 22
0: libMesh::print_trace(std::ostream&)
1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
2: void mooseError<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&)
3: Factory::reportUnregisteredError(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const
4: Factory::getValidParams(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const
5: MooseObjectAction::MooseObjectAction(InputParameters const&)
6: AddKernelAction::AddKernelAction(InputParameters const&)
7: RegistryEntry<AddKernelAction>::buildAction(InputParameters const&)
8: ActionFactory::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters&)
9: Moose::Builder::walkRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, hit::Node*)
10: Moose::Builder::walk(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, hit::Node*)
11: hit::Node::walk(hit::Walker*, hit::NodeType, hit::TraversalOrder)
12: hit::Node::walk(hit::Walker*, hit::NodeType, hit::TraversalOrder)
13: hit::Node::walk(hit::Walker*, hit::NodeType, hit::TraversalOrder)
14: Moose::Builder::build()
15: MooseApp::setupOptions()
16: MooseApp::run()
17: ./difluoride-opt(+0x30eb) [0x55fde79500eb]
18: main
19: /lib64/libc.so.6(+0x29590) [0x7f7d54a29590]
20: __libc_start_main
21: ./difluoride-opt(+0x3396) [0x55fde7950396]
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0

I tried renaming the files and application, but I didn't see any changes. Any help would be appreciated. I can provide the kernel and input file if that would be needed.

**Comment by GiudGiud** [comment]
Hello

>  I remember to change the name of the application within the .C kernel to re-register it to the new application.
All this being done, I tried to run the application and I got the following error:

- what did you register it to?
- did you remember to recompile the opt executable in between?
Guillaume

**Reply by AmbroiseJuston** [reply]
I registered it with "difluorideApp" and did recompile it in between. i also did make clear in between to ensure that it was redone.

Example code:

#include "DiffusionCoupled.h"

// Don't forget to register your object with MOOSE
registerMooseObject("difluorideApp", DiffusionCoupled);

**Reply by GiudGiud** [reply, marked as answer]
try MooseApp instead of 'difluorideApp'

**Reply by AmbroiseJuston** [reply]
i Just did and I get the exact same error message. I did try changing the name previously, to something like DiffusionCoupledCustom to see if that ould work and also changing the name of the .h.

**Reply by GiudGiud** [reply]
the file must not be compiled then

where is it in the app?

**Reply by AmbroiseJuston** [reply]
I retired it today, and changing it to MooseApp seems to register it; I might have either forgotten to save the changes or recompiled it last time in my haste. I am getting a different error now (for which I think I have the solution), so let me make sure it runs, and then I will mark it as the answer.

What is the difference between registering with " difluorideApp " and " MooseApp "? I used the previous registration scheme on previous projects, "xenonApp," and the kernels seemed to work fine there.

**Reply by GiudGiud** [reply]
The difference is that MooseApp always works, whereas if you got a single letter wrong in " difluorideApp" or if it should be " DifluorideApp" instead, then it wont work

**Reply by AmbroiseJuston** [reply]
Thank you very much. I assume a recommended practice is to write all custom kernels to MooseApp unless it is hyperspecific to the application?

**Reply by GiudGiud** [reply]
In your application, you should register the kernels to your application. But if for any reason it does not work, registering to MooseApp does not hurt anyone
